src/main/java/com/cloudbees/jenkins/plugins/bitbucket/avatars/AvatarCache.java

```diff
@@ -0,0 +1,626 @@
+                    rsp.addHeader("Cache-control", "max-age=3600, public");
+                    rsp.setStatus(HttpServletResponse.SC_NOT_MODIFIED);
+                }
+            };
+        }
+        // If no image, generate a temp avatar
+        if (avatar.image == null) {
+            // we can retry in an hour
+            return new ImageResponse(generateAvatar(avatar.source.hashKey(), targetSize), true, -1L,
+                    "max-age=3600, public");
+        }
+
+        BufferedImage image = avatar.image;
+        boolean flushImage = false;
+        if (image.getWidth() != targetSize || image.getHeight() != targetSize) {
+            image = scaleImage(image, targetSize);
+            flushImage = true;
+        }
+        return new ImageResponse(image, flushImage, avatar.lastModified, "max-age=3600, public");
+    }
+
+    /**
+     * Retrieves the entry from the cache.
+     *
+     * @param key the cache key.
+     * @param url the URL to fetch if the entry is missing or {@code null} to
+     *            perform a read-only check.
+     * @return the entry or {@code null} if a read-only check found no matching
+     *         entry.
+     */
+    @Nullable
+    private CacheEntry getCacheEntry(@NonNull final String key, @Nullable final AvatarCacheSource source) {
+        CacheEntry entry = cache.get(key);
+        if (entry == null) {
+            synchronized (serviceLock) {
+                entry = cache.get(key);
+                if (entry == null) {
+                    if (source == null) {
+                        return null;
+                    }
+                    entry = new CacheEntry(source, service.submit(new FetchImage(source)));
+                    cache.put(key, entry);
+                }
+            }
+        } else {
+            if (entry.isStale()) {
+                synchronized (serviceLock) {
+                    if (!entry.pending()) {
+                        entry.setFuture(service.submit(new FetchImage(entry.source)));
+                    }
+                }
+            }
+        }
+        entry.touch();
+        if (iterator == null || !iterator.hasNext()) {
+            synchronized (serviceLock) {
+                if (iterator == null || !iterator.hasNext()) {
+                    iterator = cache.entrySet().iterator();
+                }
+            }
+        } else {
+            synchronized (iterator) {
+                // process one entry in the cache each access
+                if (iterator.hasNext()) {
+                    Map.Entry<String, CacheEntry> next = iterator.next();
+                    if (next.getValue().isUnused()) {
+                        iterator.remove();
+                    }
+                } else {
+                    iterator = null;
+                }
+            }
+        }
+        return entry;
+    }
+
+    /**
+     * A cache entry.
+     */
+    private static class CacheEntry {
+        /**
```

Comment:
Missing white space separation
```suggestion

        /**
```